From padmanabh.ratnakar@emulex.com  Thu Jun 23 15:40:26 2011
From: Padmanabh Ratnakar <padmanabh.ratnakar@emulex.com>
Date: Tue, 10 May 2011 05:12:17 +0000
Subject: [PATCH] be2net: Handle error completion in Lancer
Git-commit: 009dd872d753f854cf13c8334e0055092f539b38
Patch-mainline: v3.0-rc1
References: FATE#311448, bnc#697255

In Lancer if a frame is DMAed partially due to lack of RX buffers,
an error completion is sent with packet size as zero and num_recvd
indicating number of used buffers. These buffers need to be freed
and packet dropped.

Signed-off-by: Padmanabh Ratnakar <padmanabh.ratnakar@emulex.com>
Signed-off-by: David S. Miller <davem@davemloft.net>

Acked-by: Jeff Mahoney <jeffm@suse.com>
---
 drivers/net/benet/be_main.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)
--- a/drivers/net/benet/be_main.c
+++ b/drivers/net/benet/be_main.c
@@ -1665,12 +1665,15 @@ static int be_poll_rx(struct napi_struct
 			break;
 
 		/* Ignore flush completions */
-		if (rxcp->num_rcvd) {
+		if (rxcp->num_rcvd && rxcp->pkt_size) {
 			if (do_gro(rxcp))
 				be_rx_compl_process_gro(adapter, rxo, rxcp);
 			else
 				be_rx_compl_process(adapter, rxo, rxcp);
+		} else if (rxcp->pkt_size == 0) {
+			be_rx_compl_discard(adapter, rxo, rxcp);
 		}
+
 		be_rx_stats_update(rxo, rxcp);
 	}
 
